### PR TITLE
Adds documentation for natural loops

### PR DESCRIPTION
**Pull request by JohnDumbell**

### CODE REVIEW

File: src/analyses/natural_loops.h [hunk 1]
```diff
@@ -7,7 +7,17 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
 \*******************************************************************/
 
 /// \file
-/// Compute natural loops in a goto_function
+/// Compute natural loops in a goto_function.
+///
+/// A natural loop is when a graphs nodes make one self-encapsulating circle
```
Comment (chrisr-diffblue) [inline src/analyses/natural_loops.h]: "when the nodes and edges of a graph"

File: src/analyses/natural_loops.h [hunk 2]
```diff
@@ -7,7 +7,17 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
...
+/// with no incoming edges from external nodes. This is to say that
```
Comment (chrisr-diffblue) [inline src/analyses/natural_loops.h]: Rather than writing "This is to say that", just say "For example....."

File: src/analyses/natural_loops.h [hunk 3]
```diff
@@ -7,7 +7,17 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
...
+/// A -> B -> C -> D -> A is a natural loop, but if B has an incoming edge from X,
+/// then it isn't a natural loop, because X is an external node. Outgoing edges don't
+/// affect the natural-ness of a loop.
+///
+/// The detection to work out if a nodes children can only be accessed through itself,
```
Comment (chrisr-diffblue) [inline src/analyses/natural_loops.h]: I might be tempted to invert the order of this paragraph a bit, and say something like "\ref cfg_dominators_templatet is used to provide the dominator analysis that is used to determine if a node is reachable only through itself, and is thus part of a natural loop." ?

File: src/analyses/natural_loops.h [hunk 4]
```diff
@@ -20,6 +30,12 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
 
 #include "cfg_dominators.h"
 
+/// Main driver for working out if a class (normally goto_programt) has any natural loops.
+/// \ref compute takes an entire goto_programt, iterates over the instructions and for
+/// any backwards jumps attempts to find out if it's a natural loop.
+///
+/// All instructions in a natural loop are stored into \ref loop_map, keyed by
+/// their header - the target of the backwards goto jump.
 template<class P, class T>
```
Comment (chrisr-diffblue) [inline src/analyses/natural_loops.h]: Might be worth elaborating what P (the program representation) and T (the node type) need to provide for there to be a valid instantiation of this template?

File: src/analyses/natural_loops.h [hunk 5]
```diff
@@ -60,6 +76,8 @@ class natural_loops_templatet
   void compute_natural_loop(T, T);
 };
 
+/// Helper class for a concretized version of
```
Comment (chrisr-diffblue) [inline src/analyses/natural_loops.h]: This isn't really a helper class, it's effectively a typedef, so maybe just call it "A concretized version of...

**Review by chrisr-diffblue — COMMENTED** [review]
Looks good in general, with a couple of small comments that I think might help improve it further.

File: src/analyses/README.md [hunk 6]
```diff
@@ -92,7 +92,17 @@ To be documented.
 
 \subsection analyses-natural-loop Natural loop analysis (natural_loops_templatet)
 
-To be documented.
+A natural loop is when a graphs nodes make one self-encapsulating circle
```
Comment (peterschrammel) [inline src/analyses/README.md]: graph's

File: src/analyses/README.md [hunk 7]
```diff
@@ -92,7 +92,17 @@ To be documented.
...
+with no incoming edges from external nodes. This is to say that
+A -> B -> C -> D -> A is a natural loop, but if B has an incoming edge from X,
+then it isn't a natural loop, because X is an external node. Outgoing edges don't
+affect the natural-ness of a loop.
+
+The detection to work out if a nodes children can only be accessed through itself,
+and find out if something is inherently natural, is called dominator analysis and
+is covered in a seperate subsection.
```
Comment (peterschrammel) [inline src/analyses/README.md]: separate

File: src/analyses/README.md (hunk 6 above)
Comment (peterschrammel) [reply src/analyses/README.md]: Provide a link to the file/class that implements that

File: src/analyses/natural_loops.h [hunk 8]
```diff
@@ -20,6 +28,20 @@ Author: Georg Weissenbacher, georg@weissenbacher.name
 
 #include "cfg_dominators.h"
 
+/// Main driver for working out if a class (normally goto_programt) has any natural loops.
+/// \ref compute takes an entire goto_programt, iterates over the instructions and for
+/// any backwards jumps attempts to find out if it's a natural loop.
+///
+/// All instructions in a natural loop are stored into \ref loop_map, keyed by
+/// their header - the target of the backwards goto jump.
```
Comment (peterschrammel) [inline src/analyses/natural_loops.h]: I'd say 'head' (i.e. loop head)

**Comment by JohnDumbell** [comment]
Applied the review comments.

**Review by peterschrammel — APPROVED** [review]
LGTM

**Pull request merged**